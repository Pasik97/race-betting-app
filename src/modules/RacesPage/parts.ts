import styled from "styled-components";

export const RacesPageWrapper = styled.div`
    position: relative;
    min-height: 150px;
`;
