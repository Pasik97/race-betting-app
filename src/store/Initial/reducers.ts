const initReducer = (state = 0, action: any) => {
   switch (action.type) {
      case 'action':
         return state + 1;
      default:
         return state;
   }
}

export default initReducer;
