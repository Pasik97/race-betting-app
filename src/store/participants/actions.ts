import { AxiosError } from 'axios';
import { Participant } from 'api/apiModels';
import * as C from './constants';

export const getParticipantsRequest = (): Extract<C.ParticipantsAction, { type: C.ParticipantsActionType.GetParticipantsRequest }> => ({
   type: C.ParticipantsActionType.GetParticipantsRequest,
});

export const getParticipantsSuccess = (participants: Participant[]): Extract<C.ParticipantsAction, { type: C.ParticipantsActionType.GetParticipantsSuccess }> => ({
   type: C.ParticipantsActionType.GetParticipantsSuccess,
   participants,
});

export const getParticipantsFail = (error: AxiosError): Extract<C.ParticipantsAction, { type: C.ParticipantsActionType.GetParticipantsFail }> => ({
   type: C.ParticipantsActionType.GetParticipantsFail,
   error,
});
