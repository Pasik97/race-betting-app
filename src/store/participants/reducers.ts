import * as C from './constants';

const participantsReducer = (state: C.ParticipantsState = C.initilParticipantsState, action: C.ParticipantsAction): C.ParticipantsState => {
   switch (action.type) {
      case C.ParticipantsActionType.GetParticipantsRequest:
         return {
            ...state,
            isFetching: true,
         }
      case C.ParticipantsActionType.GetParticipantsSuccess:
         return {
            participants: action.participants,
            isFetching: false,
         }
      case C.ParticipantsActionType.GetParticipantsFail:
         return {
            ...state,
            isFetching: false,
         }
      default:
         return state;
   }
}

export default participantsReducer;
