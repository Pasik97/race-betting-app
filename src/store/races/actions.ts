import { AxiosError } from 'axios';
import { Race } from 'api/apiModels';
import * as C from './constants';

export const getRacesRequest = (): Extract<C.RacesAction, { type: C.RaceActionType.GetRacesRequest }> => ({
   type: C.RaceActionType.GetRacesRequest,
});

export const getRacesSuccess = (races: Race[]): Extract<C.RacesAction, { type: C.RaceActionType.GetRacesSuccess }> => ({
   type: C.RaceActionType.GetRacesSuccess,
   races,
});

export const getRacesFail = (error: AxiosError): Extract<C.RacesAction, { type: C.RaceActionType.GetRacesFail }> => ({
   type: C.RaceActionType.GetRacesFail,
   error,
});
