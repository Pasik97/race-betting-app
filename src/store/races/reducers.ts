import * as C from './constants';
import { transformRacesToRacesState } from './helpers';

const racesReducer = (state: C.RacesState = C.initilRacesState, action: C.RacesAction): C.RacesState => {
   switch (action.type) {
      case C.RaceActionType.GetRacesRequest:
      case C.RaceActionType.GetRaceByIdRequest:
         return {
            ...state,
            isFetching: true,
         }
      case C.RaceActionType.GetRacesSuccess:
         return {
            ...state,
            ...transformRacesToRacesState(state.races, action.races),
            isFetching: false,
         }
      case C.RaceActionType.GetRaceByIdSuccess:
         const order = state.races[action.race.id] ? state.order : [...state.order, action.race.id];
         const races: Record<string, C.RaceWithBet> = state.races[action.race.id] ?
            { ...state.races, [action.race.id]: { ...state.races[action.race.id], ...action.race } }
            : { ...state.races, [action.race.id]: action.race };

         return {
            ...state,
            races,
            order,
            isFetching: false,
         }
      case C.RaceActionType.GetRacesFail:
      case C.RaceActionType.GetRaceByIdFail:
         return {
            ...state,
            isFetching: false,
         }
      case C.RaceActionType.SetRaceBetAmount:
         const newRaceBet: C.Bet = state.races[action.raceId]?.bet
            ? { ...state.races[action.raceId].bet, amount: action.amount }
            : { amount: action.amount };

         return {
            ...state,
            races: {
               ...state.races,
               [action.raceId]: {
                  ...state.races[action.raceId],
                  bet: newRaceBet,
               },
            },
         }
      default:
         return state;
   }
}

export default racesReducer;
