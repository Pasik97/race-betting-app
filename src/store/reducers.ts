import { combineReducers } from "redux";
import participantsReducer from "./participants/reducers";
import racesReducer from "./races/reducers";

const reducers = combineReducers({
   races: racesReducer,
   participants: participantsReducer,
})

export default reducers;
