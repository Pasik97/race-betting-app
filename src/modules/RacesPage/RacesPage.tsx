import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import * as P from './parts';
import RacesList from './RacesList/RacesList';
import { getRacesRequest } from 'store/races/actions';
import LoaderWithOverlay from 'components/LoaderWithOverlay/LoaderWithOverlay';
import { getisFetchingRaces } from 'store/races/selectors';

const RacesPage: React.FC = () => {
   const dispatch = useDispatch();

   const isFetchingRaces = useSelector(getisFetchingRaces);

   useEffect(() => {
      dispatch(getRacesRequest());
   }, [dispatch]);

   return (
      <P.RacesPageWrapper>
         <LoaderWithOverlay loading={isFetchingRaces} />
         <RacesList />
      </P.RacesPageWrapper>
   );
}

export default RacesPage;
