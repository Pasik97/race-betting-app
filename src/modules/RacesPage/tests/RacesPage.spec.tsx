
import { Provider } from "react-redux";
import { create } from "react-test-renderer";
import { createStore } from "redux";
import reducers from "store/reducers";
import RacesPage from "../RacesPage";
import * as A from 'store/races/actions';

jest.mock('modules/RacesPage/RacesList/RacesList');
jest.mock('components/LoaderWithOverlay/LoaderWithOverlay');

describe('<RacesPage />', () => {
   it('should render Loader and RacesList components', () => {
      const component = create(<Provider store={createStore(reducers)}><RacesPage /></Provider>);

      expect(component.toJSON()).toMatchSnapshot();
   });

   it('should dispatch getRacesRequest action after mount', () => {
      const getRacesRequestSpy = jest.spyOn(A, 'getRacesRequest');

      let component = create(<Provider store={createStore(reducers)}><RacesPage /></Provider>);

      component.update(<Provider store={createStore(reducers)}><RacesPage /></Provider>);

      expect(getRacesRequestSpy).toHaveBeenCalled();
   });
});
