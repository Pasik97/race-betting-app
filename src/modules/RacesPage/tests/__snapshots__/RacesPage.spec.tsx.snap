// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RacesPage /> should render Loader and RacesList components 1`] = `
<div
  className="sc-bdvvtL eEHpxK"
>
  (LoaderWithOverlay /)
  (RacesList /)
</div>
`;
